packaging/pep517_backend/_backend.py

```diff
@@ -0,0 +1,108 @@
+
+
+def _get_package_distribution_version():
+    """Retrieve the current version number from setuptools config."""
+    setup_cfg_path = Path.cwd() / 'setup.cfg'
+    setup_cfg = ConfigParser()
+    setup_cfg.read_string(setup_cfg_path.read_text())
+    cfg_version = setup_cfg.get('metadata', 'version')
+    importable_version_str = cfg_version[len('attr: '):]
+    version_mod_str, version_var_str = importable_version_str.rsplit('.', 1)
+    return getattr(import_module(version_mod_str), version_var_str)
+
+
+def _generate_rst_in_templates():
+    """Create ``*.1.rst.in`` files out of CLI Python modules."""
+    generate_man_cmd = (
+        sys.executable,
+        'hacking/build-ansible.py',
+        'generate-man',
+        '--template-file=docs/templates/man.j2',
+        '--output-dir=docs/man/man1/',
+        '--output-format=man',
+        *Path('lib/ansible/cli/').glob('*.py'),
+    )
+    subprocess.check_call(tuple(map(str, generate_man_cmd)))
+    return Path('docs/man/man1/').glob('*.1.rst.in')
+
+
+def _convert_rst_in_template_to_manpage(rst_in, version_number):
+    """Render pre-made ``*.1.rst.in`` templates into manpages.
+
+    This includes pasting the hardcoded version into the resulting files.
+    The resulting ``in``-files are wiped in the process.
+    """
+    templated_rst_doc = rst_in.with_suffix('')
+    templated_rst_doc.write_text(
+        rst_in.read_text().replace('%VERSION%', version_number))
+
+    rst_in.unlink()
+
+    rst2man_cmd = (
+        sys.executable,
+        Path(sys.executable).parent / 'rst2man.py',
+        templated_rst_doc,
+        templated_rst_doc.with_suffix(''),
+    )
+    subprocess.check_call(tuple(map(str, rst2man_cmd)))
+    templated_rst_doc.unlink()
+
+
+def build_sdist(  # noqa: WPS210, WPS430
+         sdist_directory: os.PathLike,
+         config_settings: dict[str, str] | None = None,
+) -> str:
+    build_manpages_requested = '--build-manpages' in (
+        config_settings or {}
+    )
+    if build_manpages_requested:
+        Path('docs/man/man1/').mkdir(exist_ok=True, parents=True)
+        _make_in_tree_ansible_importable()
+        version_number = _get_package_distribution_version()
+        for rst_in in _generate_rst_in_templates():
+            _convert_rst_in_template_to_manpage(rst_in, version_number)
+
+    return _setuptools_build_sdist(
+        sdist_directory=sdist_directory,
+        config_settings=config_settings,
+    )
+
+
+def get_requires_for_build_sdist(
+        config_settings: dict[str, str] | None = None,
+) -> list[str]:
+    return _setuptools_get_requires_for_build_sdist(
+        config_settings=config_settings,
+    ) + [
+        'docutils',  # provides `rst2man`
+        'jinja2',  # used in `hacking/build-ansible.py generate-man`
+        'straight.plugin',  # used in `hacking/build-ansible.py` for subcommand
+        'pyyaml',  # needed for importing in-tree `ansible-core` from `lib/`
```

Comment:
What options do we have for pinning these requirements, either here or something an end-user (and our tests) could do?

Reply:
I had originally started with the same question (and wondering if maybe they should come from our own section in `pyproject.toml`, but IIRC it's basically the same as the other build deps with PEP517- there's no official mechanism to add constraints for build deps beyond what comes from the build system. Since we're customizing it anyway now, we could technically do whatever we want by altering via these hooks, though (at least with anything past the first batch that bootstraps the build system itself, which still wouldn't be customizable).

Reply:
@mattclay setting the `PIP_CONSTRAINT` env var does the job. In fact, I've implemented this in the integration test but only included setuptools+wheel, ignoring these. It's probably a good idea to extend the pins with these. I mostly wanted to focus on going through "old" (pinned) and "latest" (unpinned) cases there.

Reply:
Testing against multiple pinned versions is fine, but we shouldn't be testing against latest since we want to avoid spontaneous test failures. These tests will need regular updates anyways for new Python versions, which will be a good time to update the pinned versions.

Reply:
Well, it's actually a bit more flexible than that and may be dependent on the build frontend and build env provisioning. `PIP_CONSTRAINT` will work for `pip` (as a build front-end, so `pip wheel ...` and `pip install --no-binary=ansible-core ...` — the env var is passed all the way down to the tmp build venv, unlike the `--constraint` CLI option) and for the default provisioning mode of `pypa/build` which uses pip for populating the tmp build venv.

But `pypa/build` has flags to skip the provisioning of the isolated build env as well as installing the build deps. It is intended for more granular build deps control. The downstreams could use this, for example. Though, I know that downstreams have helpers like https://src.fedoraproject.org/rpms/pyproject-rpm-macros (Fedora) and https://pypi.org/p/gpep517 (Gentoo).

And, of course, there's libs implementing PEP 517. There's old https://pypi.org/p/pep517 package and its new replacement — https://pypi.org/p/pyproject_hooks. The ecosystem is currently migrating to the latter — it's intended to cover more than just PEP 517 attempting to have a more generic name, including more PEPs over time. The last `pypa/build` version is already using it.
FWIW, those who'd want even more flexibility would use it to extract the deps.

Oh, and pip-tools also has a PR in the works, facilitating pinning said build deps: https://github.com/jazzband/pip-tools/pull/1681.

Reply:
> I wonder if we could still smoke-test the bleeding-edge deps somehow 

We've had this dream for years (eg a nightly/weekly "what's busted with all the upper version limit constraints removed?" full CI run), but we've never had time to go implement it. It's arguably getting easier as we're using pinned constraints more and the various dep installs have been getting more centralized, but it's still not a super-easy thing to do. :(

Reply:
> I had originally started with the same question (and wondering if maybe they should come from our own section in `pyproject.toml`, but IIRC it's basically the same as the other build deps with PEP517- there's no official mechanism to add constraints for build deps beyond what comes from the build system. Since we're customizing the build anyway now, though, we could technically do whatever we want by altering via these hooks (at least with anything past the first batch that bootstraps the build system itself, which still wouldn't be customizable).

Just noticed this comment. I've got some points to add:
1) Bootstrapping the build system itself could also go into the hook if we *really wanted that* — it'd just need more try/excepts around imports; though having them in a TOML file does _seem_ cleaner
2) It is important to recognize that the `get_requires_for_build_sdist` hook only augments the deps required for building sdist but not wheel, while `[build-system] requires` is used for both cases. By the way, setuptools' `get_requires_for_build_wheel` adds the dependency on `wheel` similarly.
3) If we were to follow PEP 518 to list the additional deps in `pyproject.toml`, it'd require adding even more deps, like `tomlkit` (below Python 3.11) — just to be able to read said config :man_shrugging:
4) In the PR implementing https://github.com/jazzband/pip-tools/issues/1396, I've suggested using the standardized hooks to get lists of the direct deps for different actions. So I think it's not entirely “no official mechanism” — if we were to put the deps in `pyproject.toml`, the tooling would still have to go through the hooks to retrieve the deps (and those hooks would read the config) and not parse TOML directly.

Reply:
> > I wonder if we could still smoke-test the bleeding-edge deps somehow
> 
> We've had this dream for years (eg a nightly/weekly "what's busted with all the upper version limit constraints removed?" full CI run), but we've never had time to go implement it. It's arguably getting easier as we're using pinned constraints more and the various dep installs have been getting more centralized, but it's still not a super-easy thing to do. :(

I've made some automation in Cheroot that runs pip-tools periodically and updates the PR. I wanted to extend that to keep a separate "stable-pins-bump" branch with passing tests and have an additional "unstable-pins" branch for bumps. If the update to "unstable-pins" results in a passing CI, it'd fast-forward the stable one. This would allow making version bump merges less often...
But yeah, it can be a lot of work still.